Enforce expiry on eq_session

MebinAbraham opened the pull request:
### What is the context of this PR?
We do not need to keep all the session data in DynamoDB. If a session hasn't been updated for a set period the session is no longer valid and can be deleted from our table. Currently session is not updated per request hence all users are signed out after the preset 45 minutes.

Issues being addressed:
- `updated_at` time not being updated hence session times out after 45 minutes (current timeout length), regardless of whether you were active or not.
- browser cookie never expires - current expiry time `(1969-12-31T23:59:59.000Z)`

This PR introduces a new column `expires_at` which is used to determine whether a user has a valid session. This column has been introduced to both Postgres and Dynamo to keep the data consistent. For Postgres, a DB migration is required, which will be done through another PR. 

Once the `expires_at` time has passed, the user will be signed out when making the next request, or automatically via the timeout modal (JS) feature. This value is also used to configure [TTL](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html) within DynamoDB, which will delete the data within 48hrs of expiry.

---

To test DynamoDB, provision Dynamo using eq-terraform using the branch `eq-session-ttl-enabled`. Make sure you change the env variable in runner to match your environment eg 
`dev-eq-session` - > `<name>-eq-session`
For context here are my current env settings: https://gist.github.com/MebinAbraham/11160b4daa9163aec1b65ee9cdc46d9f
Toggle: `EQ_SESSION_DYNAMO_WRITE=True` to test between Dynamo/SQLite

To test the changes for Postgres (SQLite locally), you need to alter the table manually. Use contents from PR #1797 to add a new column in SQLite.
`cat database-scripts/pr-1797-apply.sql | sqlite3 /tmp/questionnaire.db`
or just remove the tmp db. `rm /tmp/questionnaire.db`

You can test the expiry by using the `test_timeout.json` schema or by changing these in `settings.py`:
```
EQ_SESSION_TIMEOUT_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_SECONDS', str(45 * 60)))
EQ_SESSION_TIMEOUT_GRACE_PERIOD_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_GRACE_PERIOD_SECONDS', '5'))
EQ_SESSION_TIMEOUT_PROMPT_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_PROMPT_SECONDS', '5'))
```

### How to review 
Test all scenarios that you can think of.
- Ensure user is signed out when their session has expired.
- Ensure the `updated_at` and `expires_at` time is modified with each request.
- Use chrome dev tool to inspect the session cookie to ensure the expiry time is as expected. `(Inspect Element > Application > Cookies > http://localhost:5000 > session > Expires/Max-Age)`

### Expected behaviour
- All new sessions are given an expiration time.
- Once the expiration time has passed, the user is signed out.
- In-flight sessions will not be given an expiration time. (This can be discussed, currently, it won't since the `expires_at` is retrieved from the cookie).
- All requests should be backwards compatible.

Note: I have still have some test to write for the different scenarios but wanted to get some eyes on this.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

jonnyshaw89 commented on app/authentication/authenticator.py:
```diff
@@ -40,23 +41,57 @@ def when_user_logged_out(sender_app, user):  # pylint: disable=unused-argument
+        session_store.expiration_time = datetime.utcnow() + timedelta(seconds=session_timeout)
+        session_store.save()
+
+        logger.info('session expiry extended')
```
This is going to get a bit loggy?

jonnyshaw89 commented on app/helpers/template_helper.py:
```diff
@@ -38,6 +33,10 @@ def with_metadata_context(func):
         metadata = get_metadata(current_user)
+
+        if not metadata:
+            raise Unauthorized
```
Is this not handled elsewhere?
can you get this far without having metadata?

MebinAbraham replied:
I added this in when going to feedback page without a session threw a 500, but it seems to magically work now. I'll remove.

jonnyshaw89 commented on app/settings.py:
```diff
@@ -51,9 +51,9 @@ def get_env_or_fail(key):
-EQ_SESSION_TIMEOUT_PROMPT_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_PROMPT_SECONDS', '120'))
+EQ_SESSION_TIMEOUT_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_SECONDS', str(10)))
+EQ_SESSION_TIMEOUT_GRACE_PERIOD_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_GRACE_PERIOD_SECONDS', '5'))
+EQ_SESSION_TIMEOUT_PROMPT_SECONDS = int(os.getenv('EQ_SESSION_TIMEOUT_PROMPT_SECONDS', '5'))
```
Are these changes just you testing?

MebinAbraham replied:
Yes, will revert before.

jonnyshaw89 replied:
why not now?

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -454,6 +449,16 @@ def _store_viewable_submission(answers, metadata, submitted_time):
+#     _eq_session.expires_at = datetime.utcnow() + timedelta(seconds=get_session_timeout_in_seconds(g.schema))
+#
+#     data_access.put(_eq_session)
+
```
This is elsewhere now so can go?

bitdivision commented on app/authentication/authenticator.py:
```diff
@@ -40,23 +41,57 @@ def when_user_logged_out(sender_app, user):  # pylint: disable=unused-argument
+
         return user
 
+    if session_store:
```
You're checking this in `_is_valid_session` too. 

I don't know why, but I found the control flow here a little bit difficult to grok. Personally I would go for:
```
if not session_store:
    return None

if _is_valid_session(session_store) and session_store.user_id:
    ...
    return user
else:
    session_store.delete()
    ...
```

MebinAbraham replied:
Fair point. Updated. Didn't use `else` because of linting ` Unnecessary "else" after "return" (no-else-return)`.

bitdivision reviewed (commented):
Haven't done any testing yet, but code looks good so far.

insacuri commented on app/data_model/app_models.py:
```diff
@@ -1,4 +1,4 @@
-import datetime
+from datetime import datetime
```
👍

insacuri commented on app/data_model/session_store.py:
```diff
@@ -1,5 +1,5 @@
-from structlog import get_logger
 import simplejson as json
+from structlog import get_logger
```
Not used now so this can be removed

insacuri replied:
Ignore that sorry! I thought this was being added and I couldn't see any new log lines so I thought it wasn't being used.

insacuri commented:
Code looks good but I've not tested it yet.

jonnyshaw89 commented:
Ive tested expiring and deleting the token from the DB and it kicks me out 👍

MebinAbraham commented:
The expiry and updated time can also be seen updating in Dynamo/SQLite when you run the functional tests.

MebinAbraham commented:
Okay I broke it for in-flight sessions for Dynamo. 

Edit: This should never happen but added a fix anyway.

jonnyshaw89 commented on app/data_model/models.py:
```diff
@@ -40,22 +42,25 @@ class EQSession(db.Model):
     session_data = db.Column('session_data', db.String)
     created_at = db.Column('created_at', db.DateTime, default=datetime.datetime.utcnow)
-    updated_at = db.Column('updated_at', db.DateTime, default=datetime.datetime.utcnow, onupdate=datetime.datetime.utcnow)
+    updated_at = db.Column('updated_at', db.DateTime)
```
How does the `updated_at` get updated now?

bitdivision replied:
In `__init__()`

jonnyshaw89 replied:
so we set updated when we load from the DB?

bitdivision replied:
Indeed. I was a little skeptical too.

MebinAbraham replied:
Trying to keep it similar, but I can also do `on_update` since `expires_at` would be updated.

ajmaddaford commented on app/globals.py:
```diff
@@ -42,12 +44,35 @@ def get_session_store():
+
+    return schema_session_timeout if schema_session_timeout and \
+                                     schema_session_timeout < default_session_timeout else \
+                                     default_session_timeout
```
It's confusing to add the grace period in two places, this would read more easily as:

```
default_session_timeout = current_app.config['EQ_SESSION_TIMEOUT_SECONDS']
schema_session_timeout = schema.json.get('session_timeout_in_seconds')
timeout = schema_session_timeout if schema_session_timeout and \
          schema_session_timeout < default_session_timeout else \
          default_session_timeout

if with_grace_period:
    grace_period = current_app.config['EQ_SESSION_TIMEOUT_GRACE_PERIOD_SECONDS']
    timeout += grace_period

return timeout
```

MebinAbraham replied:
Sweet, will change.

ajmaddaford commented on app/authentication/authenticator.py:
```diff
@@ -40,24 +41,53 @@ def when_user_logged_out(sender_app, user):  # pylint: disable=unused-argument
+    """
+    checks that the user has a valid session
+    """
+    if session_store and session_store.user_id:
```
When is `user_id` not set on `session_store`?

ajmaddaford commented on app/authentication/authenticator.py:
```diff
@@ -40,24 +41,53 @@ def when_user_logged_out(sender_app, user):  # pylint: disable=unused-argument
 
+def _extend_session_expiry(session_store):
+    session_timeout = cookie_session.get('expires_in')
+    if session_store and session_store.expiration_time and session_timeout:
```
Is the check for `session_store.expiration_time` only required whilst we have in flight sessions? If so it may be worth highlighting it so we know to remove it.

MebinAbraham replied:
Yes it is. Will add a comment.

ajmaddaford commented on app/data_model/app_models.py:
```diff
@@ -9,17 +9,21 @@ def __init__(self, user_id, state_data, version):
-        self.created_at = datetime.datetime.now(tz=tzutc())
-        self.updated_at = datetime.datetime.now(tz=tzutc())
+        self.created_at = datetime.now(tzutc())
+        self.updated_at = datetime.now(tzutc())
```
These are inconsistent with the `created_at` and `updated_at` above that use `tz=tzutc()`

bitdivision commented on tests/app/data_model/test_app_models.py:
```diff
@@ -16,14 +15,17 @@ def test_submitted_response(self):
-        self.assertGreaterEqual(new.updated_at, NOW)
+        new_model = self._test_model(QuestionnaireState('someuser', 'somedata', 1))
+
+        self.assertGreaterEqual(new_model.created_at, NOW)
```
We do have `freezegun` in the dependencies now, which you could use here. Not particularly important in this case, but worth bearing in mind if you're writing these kinds of tests.

MebinAbraham replied:
Will have a look into it. Don't see why we have it in the dependencies since we don't seem to use it anywhere currently?.

bitdivision replied:
We use it in `test_date_reference_ranges.py`

bitdivision replied:
https://github.com/ONSdigital/eq-survey-runner/blob/master/tests/integration/mci/test_date_reference_ranges.py#L4

bitdivision commented:
I've tested the basics, and I'm happy with the code. 
I'm still getting weird terraform errors, which seem to be particular to my env. I'll try to resolve those tomorrow and get this merged.

dcdarrell9 reviewed (commented):
LGTM 👍  Had no issues

MebinAbraham commented:
@ajmaddaford Have made the changes and ready for re-review.
@jonnyshaw89 Could you make sure you are still happy with the new changes.
Thanks

ajmaddaford reviewed (approved):
Looks good. Tested in both Postgres and DynamoDB in AWS.

Commit:
Enforced expiration on eq_session table

 - New expires_at column stores session expiration time
 - Session expiry and last updated time is now updated on every request
 - Set client-side cookie expiration
